### PR TITLE
Core/Scripts: Hallow's End Pumpkin Treat

### LINKED ISSUE #3573: Hallows End [item] Hallow's End Pumpkin Treat 20557
Missing the 4 possible auras when eaten
    Great Pumpkin 24924 which should have a chance trigger Hallow's End Fright 24924 (can proc a chill/ slow effect) on enemy
    Skeletal Giant 24925 
    Ghost 24927 
    Pirate  24926 - no model change or swim speed increase

up 42
dc2bdb8

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/66133-hallows-end-item-hallow-s-end-pumpkin-treat-20557?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ghost**
**Changes proposed**:

Make all 6 spells and their auras available from [item 20557, Hallow's End Pumpkin Treat](http://www.wowhead.com/item=20557/hallows-end-pumpkin-treat):
- [Hallow's End Candy, Orange Giant (24924)](http://www.wowhead.com/spell=24924/hallows-end-candy)
- [Hallow's End Candy, Skeleton (24925)](http://www.wowhead.com/spell=24925/hallows-end-candy)
- [Hallow's End Candy, Pirate (24926)](http://www.wowhead.com/spell=24926/hallows-end-candy)
- [Hallow's End Candy, Ghost (24927)](http://www.wowhead.com/spell=24927/hallows-end-candy)
- [Hallow's End Candy, Female Defias Pirate (44742)](http://www.wowhead.com/spell=44742/hallows-end-candy-female-dnd)
- [Hallow's End Candy, Male Defias Pirate (44743)](http://www.wowhead.com/spell=44743/hallows-end-candy-male-dnd)

~~The core script is based on suggestions from @KoffeinKaio and @Anubisss in #3573~~

Thanks @Keader for creating the new spell scripts & SQL, saved me a lot of time! :-)

**Target branch(es)**: 3.3.5 (don't know if this is available in 6.x)

**Issues addressed**: Closes #3573 **[Hallows End [item] Hallow's End Pumpkin Treat 20557](3573)**

**Tests performed**: (Does it build, tested in-game, etc)

Builds without any issues, tested and found to be working in-game.

**Known issues and TODO list**:
- [x] Possible visual model issue with the spells 44742 + 44743 if game client Cache is not deleted.

### CODE REVIEW

**Comment by P-Kito** [comment]
What do you guys think of:

```C++
enum SpellsPumpkinTreat // Item 20557, Hallow's End Pumpkin Treat
{
    SPELL_HALLOWS_END_CANDY_ORANGE_GIANT         = 24924, // Effect 1: Apply Aura: Mod Size, Value: 30%
    SPELL_HALLOWS_END_CANDY_SKELETON             = 24925, // Effect 1: Apply Aura: Change Model (Skeleton), 2) Apply Aura: Underwater Breathing
    SPELL_HALLOWS_END_CANDY_PIRATE               = 24926, // Effect 1: Apply Aura: Increase Swim Speed, Value: 50%
    SPELL_HALLOWS_END_CANDY_GHOST                = 24927, // Effect 1: Apply Aura: Levitate / Hover, 2: Apply Aura: Slow Fall, 3: Apply Aura: Water Walking
    SPELL_HALLOWS_END_CANDY_FEMALE_DEFIAS_PIRATE = 44742, // Effect 1: Apply Aura: Change Model (Defias Pirate, Female), 2: Increase Swim Speed, Value: 50%
    SPELL_HALLOWS_END_CANDY_MALE_DEFIAS_PIRATE   = 44743  // Effect 1: Apply Aura: Change Model (Defias Pirate), 2: Increase Swim Speed, Value: 50%
};

uint32[] spells = { 
SPELL_HALLOWS_END_CANDY_ORANGE_GIANT, SPELL_HALLOWS_END_CANDY_SKELETON,
SPELL_HALLOWS_END_CANDY_PIRATE, SPELL_HALLOWS_END_CANDY_GHOST,
SPELL_HALLOWS_END_CANDY_FEMALE_DEFIAS_PIRATE, SPELL_HALLOWS_END_CANDY_MALE_DEFIAS_PIRATE };

And then in function replace the whole 
`switch (urand(0,5))` 
with 
`player->CastSpell(player, spells[urand(0,5)], true);`

Would this be better?
And do you have to check `HasAura()` for `RemoveAura()`?

File: src/server/scripts/World/item_scripts.cpp [hunk 1]
```diff
@@ -425,4 +489,5 @@ void AddSC_item_scripts()
     new item_dehta_trap_smasher();
     new item_trident_of_nazjan();
     new item_captured_frog();
+	new item_pumpkin_treat();
```
Comment (P-Kito) [inline src/server/scripts/World/item_scripts.cpp]: Whitespace :)

File: src/server/scripts/World/item_scripts.cpp [hunk 2]
```diff
@@ -412,6 +412,70 @@ class item_captured_frog : public ItemScript
...
+                    player->CastSpell(player, SPELL_HALLOWS_END_CANDY_FEMALE_DEFIAS_PIRATE, true);
+                    break;
+                case 5:
+                    player->CastSpell(player, SPELL_HALLOWS_END_CANDY_MALE_DEFIAS_PIRATE, true);
+                    break;
+                default:
+                    break;
+            }
+
+           return false;
```
Comment (P-Kito) [inline src/server/scripts/World/item_scripts.cpp]: return true;

File: src/server/scripts/World/item_scripts.cpp (hunk 1 above)
Comment (tkrokli) [inline src/server/scripts/World/item_scripts.cpp]: Tab, in fact. :-P Will replace it with 4 spaces ASAP.

**Comment by GitVerge** [comment]
@P-Kito About HasAura() for RemoveAura(): https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Unit/Unit.cpp#L3707
In theory, when equal_range finds "nothing", range.first == range.second. Won't pass the loop condition.

**Comment by joschiwald** [comment]
Script the spell on the item instead of the item itself

HasAura is useless if you are removing auras

there is a RAND() function that return one random item of your input args

File: src/server/scripts/World/item_scripts.cpp [hunk 3]
```diff
@@ -412,6 +412,70 @@ class item_captured_frog : public ItemScript
...
+class item_pumpkin_treat : public ItemScript
+{
+    public:
+
+        item_pumpkin_treat() : ItemScript("item_pumpkin_treat") { }
+
+        bool OnUse(Player* player, Item* /*item*/, SpellCastTargets const& /*targets*/)
+        {
+            if (player->HasAura(SPELL_HALLOWS_END_CANDY_ORANGE_GIANT))
+                player->RemoveAura(SPELL_HALLOWS_END_CANDY_ORANGE_GIANT);
```
Comment (Keader) [inline src/server/scripts/World/item_scripts.cpp]: SPELL_HALLOWS_END_CANDY_ORANGE_GIANT is not "Aura" object, is a enum right?
Why not use ```player-RemoveAurasDueToSpell(SPELL_HALLOWS_END_CANDY_ORANGE_GIANT);```

**Comment by tkrokli** [comment]
OK, might as well take the time I need to write a spellscript based on existing templates. After all, this issue has been open since 2011.

**Comment by tkrokli** [comment]
Script rewrite by Keader, I have replaced my files with his suggestions.

**Comment by Keader** [comment]
if someone has sniffs of this, can confirm spell ids?

**Comment by Aokromes** [comment]
@joschiwald this can be merged?

**Comment by joschiwald** [comment]
cee1c0789acb7d75e453ed355a5d68781a5dec4c

**Pull request closed without merging**